### PR TITLE
DRAFT: Issue 4329. Handling port 0 correctly

### LINKED ISSUE #4329: Allow zero port in config
**Is your feature request related to a problem? Please describe.**
When testing Temporal, especially trying to test locally with multiple servers, it's painful to have to update the config with new free ports each time. 

**Describe the solution you'd like**
Allow users to specify 0 for any port in the `config.Config` to let the OS choose a port when we bind. As soon as that port is bound, overwrite the port variable in the config to be the actual port.

**Describe alternatives you've considered**
Currently, temporalite uses [freeport.go] to do this, but there's actually a race condition there because another process may bind to the "free" ports after the ephemeral listeners close, but before the temporalite server binds to them.

**Additional context**

[freeport.go]: https://github.com/temporalio/temporalite/blob/main/internal/liteconfig/freeport.go

### PR DESCRIPTION
**Pull request by edmondop**
**What changed?**
Fixes #4329 

**Why?**

When specifying port 0 in Unix like system, the TCP implementation looks for the first available port. However, we never update the configuration to reflect that we actually bound to that port

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

**Comment by edmondop** [comment]
In the list here https://github.com/temporalio/temporal/blob/master/common/resource/fx.go#L126 such a module appears at the end, but I am not sure that FX guarantees the order of execution of setup , so I added an explicit deps

File: common/rpc/test/rpc_common_test.go [hunk 1]
```diff
@@ -100,6 +100,7 @@ func startHelloWorldServer(s *suite.Suite, factory *TestFactory) (*grpc.Server,
 	helloworld.RegisterGreeterServer(server, greeter)
 
 	listener := factory.GetGRPCListener()
+	s.NotEqual(rpcTestCfgDefault.GRPCPort, 0)
```
Comment (MichaelSnowden) [inline common/rpc/test/rpc_common_test.go:103]: To test this properly, you'll need to add a test which sets the port to 0, starts the server, and then verifies that it is set to a non-zero number

File: common/config/fx.go [hunk 2]
```diff
@@ -49,7 +50,7 @@ func provideMembershipConfig(cfg *Config) *Membership {
 	return &cfg.Global.Membership
 }
 
-func provideServicePortMap(cfg *Config) ServicePortMap {
+func provideServicePortMap(cfg *Config, _ *net.Listener) ServicePortMap {
```
Comment (MichaelSnowden) [inline common/config/fx.go:53]: I don't think you need this extra param. It's also what's causing the tests to fail

Comment (edmondop) [reply common/config/fx.go:53]: How do you think we can enforce this function to be called only after the listener is created? It could read the port before it is set to its actual value.

I checked the documentation of `fx` and it doesn't seem to guarantee an order for constructors that have no dependencies

File: common/rpc/test/rpc_common_test.go (hunk 1 above)
Comment (edmondop) [reply common/rpc/test/rpc_common_test.go:103]: This is using 0 (see https://github.com/temporalio/temporal/pull/4411/files/9012b7924b841a6ea351e734ada3296aff35d40a#diff-8f16f53e01a0670b51e43596ee903cd2dd83e51399714ff96c0e5da69e4d3448R76)

but I understand we might want to separate the unit test for the purpose. Where should this unit test reside?

File: common/config/fx.go (hunk 2 above)
Comment (MichaelSnowden) [reply common/config/fx.go:53]: Oh, I see. I don't know if it's going to be that simple. I think the best approach would be to probably refactor this so that we aren't reading raw port values. We don't want to do any side-effects during graph building time, only during the startup lifecycle, so we can pass around functions to read the ports, and then only actually read them in the start methods of objects. It will probably be a much larger change

**Comment by edmondop** [comment]
At the moment this fails when I try to link `temporal server start-dev` to my local checkout of the Temporal server.

```
Error: could not build arguments for function "go.temporal.io/server/temporal".ServerLifetimeHooks
        /Users/edmondoporcu/Development/open-source/temporal/temporal/fx.go:848:
failed to build *temporal.ServerImpl:
could not build arguments for function "go.temporal.io/server/temporal".NewServerFxImpl
        /Users/edmondoporcu/Development/open-source/temporal/temporal/server_impl.go:60:
could not build value group *temporal.ServicesMetadata[group="services"]:
received non-nil error from function "go.temporal.io/server/temporal".HistoryServiceProvider
        /Users/edmondoporcu/Development/open-source/temporal/temporal/fx.go:365:
cycle detected in dependency graph:
func(fx.Lifecycle, ringpop.factoryParams) (membership.Monitor, error) provided by "go.temporal.io/server/common/membership/ringpop".provideMonitor (/Users/edmondoporcu/Development/open-source/temporal/common/membership/ringpop/fx.go:39)
        depends on func(*config.Config, net.Listener) config.ServicePortMap provided by "go.temporal.io/server/common/config".provideServicePortMap (/Users/edmondoporcu/Development/open-source/temporal/common/config/fx.go:53)
        depends on func(common.RPCFactory) net.Listener provided by "go.temporal.io/server/common/resource".GrpcListenerProvider (/Users/edmondoporcu/Development/open-source/temporal/common/resource/fx.go:153)
        depends on func(*config.Config, primitives.ServiceName, log.Logger, encryption.TLSConfigProvider, membership.GRPCResolver, telemetry.ClientTraceInterceptor) (common.RPCFactory, error) provided by "go.temporal.io/server/common/resource".RPCFactoryProvider (/Users/edmondoporcu/Development/open-source/temporal/common/resource/fx.go:379)
        depends on func(membership.Monitor) membership.GRPCResolver provided by "go.temporal.io/server/common/membership".initializeBuilder (/Users/edmondoporcu/Development/open-source/temporal/common/membership/grpc_resolver.go:59)
        depends on func(fx.Lifecycle, ringpop.factoryParams) (membership.Monitor, error) provided by "go.temporal.io/server/common/membership/ringpop".provideMonitor (/Users/edmondoporcu/Development/open-source/temporal/common/membership/ringpop/fx.go:39)
('export TEMPORAL_CLI_SHOW_STACKS=1' to see stack traces)
Exiting.
```

However, if I remove the explicit dependency, then the serviceMap access the port before it's bound :(
